localstack-core/localstack/services/sns/v2/provider.py

```diff
@@ -118,6 +164,360 @@ def set_topic_attributes(
+                    sub_attributes["RawMessageDelivery"] = raw_msg_delivery.lower()
+
+        # An endpoint may only be subscribed to a topic once. Subsequent
+        # subscribe calls do nothing (subscribe is idempotent), except if its attributes are different.
+        for existing_topic_subscription in store.topic_subscriptions.get(topic_arn, []):
+            sub = store.subscriptions.get(existing_topic_subscription, {})
+            if sub.get("Endpoint") == endpoint:
+                if sub_attributes:
+                    # validate the subscription attributes aren't different
+                    for attr in sns_constants.VALID_SUBSCRIPTION_ATTR_NAME:
+                        # if a new attribute is present and different from an existent one, raise
+                        if (new_attr := sub_attributes.get(attr)) and sub.get(attr) != new_attr:
+                            raise InvalidParameterException(
+                                "Invalid parameter: Attributes Reason: Subscription already exists with different attributes"
+                            )
+
+                return SubscribeResponse(SubscriptionArn=sub["SubscriptionArn"])
+        principal = DUMMY_SUBSCRIPTION_PRINCIPAL.format(
+            partition=get_partition(context.region), account_id=context.account_id
+        )
+        subscription_arn = create_subscription_arn(topic_arn)
+        subscription = SnsSubscription(
+            # http://docs.aws.amazon.com/cli/latest/reference/sns/get-subscription-attributes.html
+            TopicArn=topic_arn,
+            Endpoint=endpoint,
+            Protocol=protocol,
+            SubscriptionArn=subscription_arn,
+            PendingConfirmation="true",
+            Owner=context.account_id,
+            RawMessageDelivery="false",  # default value, will be overridden if set
+            FilterPolicyScope="MessageAttributes",  # default value, will be overridden if set
+            SubscriptionPrincipal=principal,  # dummy value, could be fetched with a call to STS?
+        )
+        if sub_attributes:
+            subscription.update(sub_attributes)
+            if "FilterPolicy" in sub_attributes:
+                filter_policy = (
+                    json.loads(sub_attributes["FilterPolicy"])
+                    if sub_attributes["FilterPolicy"]
+                    else None
+                )
+                if filter_policy:
+                    validator = FilterPolicyValidator(
+                        scope=subscription.get("FilterPolicyScope", "MessageAttributes"),
+                        is_subscribe_call=True,
+                    )
+                    validator.validate_filter_policy(filter_policy)
+
+                store.subscription_filter_policy[subscription_arn] = filter_policy
+
+        store.subscriptions[subscription_arn] = subscription
+
+        topic_subscriptions = store.topic_subscriptions.setdefault(topic_arn, [])
+        topic_subscriptions.append(subscription_arn)
```

Comment:
linked to the comment above, the first line here can be removed, and the `topic_subscriptions` will be defined earlier